File: spring/src/main/java/cardsproject/domain/tournaments/Game.java
package cardsproject.domain.tournaments;

import jakarta.persistence.*;
import jakarta.validation.constraints.*;

@Entity
@Table(name = "games")
public class Game {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Min(1)
    @Max(3)
    private Integer gameNumber = 0;
    @Enumerated(EnumType.STRING)
    private GameWinnerSideType winnerSide;
    private Double complexityScore;
    private Integer turnsPlayed;
    private Integer durationSeconds;
    @Enumerated(EnumType.STRING)
    private GameEndedByType endedBy;
    private String replayUrl;

    // @ManyToOne -> Match, onDelete=CASCADE, relatedName=games
    @Column(name = "match_id")
    private Long matchId;
    // @ManyToOne -> Player, onDelete=SET_NULL, relatedName=won_games, via=players
    @Column(name = "winner_id")
    private Long winnerId;

    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }
    public Integer getGameNumber() { return gameNumber; }
    public void setGameNumber(Integer gameNumber) { this.gameNumber = gameNumber; }
    public GameWinnerSideType getWinnerSide() { return winnerSide; }
    public void setWinnerSide(GameWinnerSideType winnerSide) { this.winnerSide = winnerSide; }
    public Double getComplexityScore() { return complexityScore; }
    public void setComplexityScore(Double complexityScore) { this.complexityScore = complexityScore; }
    public Integer getTurnsPlayed() { return turnsPlayed; }
    public void setTurnsPlayed(Integer turnsPlayed) { this.turnsPlayed = turnsPlayed; }
    public Integer getDurationSeconds() { return durationSeconds; }
    public void setDurationSeconds(Integer durationSeconds) { this.durationSeconds = durationSeconds; }
    public GameEndedByType getEndedBy() { return endedBy; }
    public void setEndedBy(GameEndedByType endedBy) { this.endedBy = endedBy; }
    public String getReplayUrl() { return replayUrl; }
    public void setReplayUrl(String replayUrl) { this.replayUrl = replayUrl; }
    public Long getMatchId() { return matchId; }
    public void setMatchId(Long matchId) { this.matchId = matchId; }
    public Long getWinnerId() { return winnerId; }
    public void setWinnerId(Long winnerId) { this.winnerId = winnerId; }

    // ── Business operations ──────────────────────────────────────────
    public void recordWinner(String winnerSide) {
        // TODO: implement recordWinner
    }
    @com.fasterxml.jackson.annotation.JsonIgnore
    public java.math.BigDecimal durationMinutes() {
        // TODO: implement durationMinutes
        return null;
    }

    // ── Validation rules ─────────────────────────────────────────────
    @jakarta.validation.constraints.AssertTrue(message = "Game number must be between 1 and 3 (best-of-3)")
    @com.fasterxml.jackson.annotation.JsonIgnore
    public boolean isGameNumberRangeValid() {
        return (getGameNumber() == null || (getGameNumber() >= 1 && getGameNumber() <= 3));
    }
}

File: symfony/src/Entity/Tournaments/Game.php
<?php

namespace App\Entity\Tournaments;

use Doctrine\ORM\Mapping as ORM;
use Symfony\Component\Serializer\Annotation\Groups;
use App\Repository\Tournaments\GameRepository;
use App\Entity\Players\Player;

#[ORM\Entity(repositoryClass: GameRepository::class)]
#[ORM\Table(name: 'game')]
class Game
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column]
    #[Groups(['game:read'])]
    private ?int $id = null;

    #[ORM\Column(type: 'integer')]
    #[Groups(['game:read', 'game:write'])]
    private int $gameNumber = 0;

    #[ORM\Column(type: 'string', length: 20, nullable: true)]
    #[Groups(['game:read', 'game:write'])]
    private ?string $winnerSide = null;

    #[ORM\Column(type: 'integer', nullable: true)]
    #[Groups(['game:read', 'game:write'])]
    private ?int $turnsPlayed = null;

    #[ORM\Column(type: 'integer', nullable: true)]
    #[Groups(['game:read', 'game:write'])]
    private ?int $durationSeconds = null;

    #[ORM\Column(type: 'string', length: 20, nullable: true)]
    #[Groups(['game:read', 'game:write'])]
    private ?string $endedBy = null;

    #[ORM\Column(type: 'string', length: 200, nullable: true)]
    #[Groups(['game:read', 'game:write'])]
    private ?string $replayUrl = null;

    #[ORM\ManyToOne(targetEntity: MatchRecord::class, inversedBy: 'games')]
    #[ORM\JoinColumn(nullable: false)]
    private ?MatchRecord $match = null;

    #[ORM\ManyToOne(targetEntity: Player::class, inversedBy: 'won_games')]
    #[ORM\JoinColumn(nullable: true)]
    private ?Player $winner = null;

    public function getId(): ?int
    {
        return $this->id;
    }

    public function getGameNumber(): int
    {
        return $this->gameNumber;
    }

    public function setGameNumber(int $gameNumber): static
    {
        $this->gameNumber = $gameNumber;
        return $this;
    }

    public function getWinnerSide(): ?string
    {
        return $this->winnerSide;
    }

    public function setWinnerSide(?string $winnerSide): static
    {
        $this->winnerSide = $winnerSide;
        return $this;
    }

    public function getTurnsPlayed(): ?int
    {
        return $this->turnsPlayed;
    }

    public function setTurnsPlayed(?int $turnsPlayed): static
    {
        $this->turnsPlayed = $turnsPlayed;
        return $this;
    }

    public function getDurationSeconds(): ?int
    {
        return $this->durationSeconds;
    }

    public function setDurationSeconds(?int $durationSeconds): static
    {
        $this->durationSeconds = $durationSeconds;
        return $this;
    }

    public function getEndedBy(): ?string
    {
        return $this->endedBy;
    }

    public function setEndedBy(?string $endedBy): static
    {
        $this->endedBy = $endedBy;
        return $this;
    }

    public function getReplayUrl(): ?string
    {
        return $this->replayUrl;
    }

    public function setReplayUrl(?string $replayUrl): static
    {
        $this->replayUrl = $replayUrl;
        return $this;
    }

    #[Groups(['game:read'])]
    public function getMatchId(): ?int
    {
        return $this->match?->getId();
    }

    public function getMatch(): ?MatchRecord
    {
        return $this->match;
    }

    public function setMatch(?MatchRecord $match): static
    {
        $this->match = $match;
        return $this;
    }

    #[Groups(['game:read'])]
    public function getWinnerId(): ?int
    {
        return $this->winner?->getId();
    }

    public function getWinner(): ?Player
    {
        return $this->winner;
    }

    public function setWinner(?Player $winner): static
    {
        $this->winner = $winner;
        return $this;
    }

    // ── Validation rules ─────────────────────────────────────────────
    #[\Symfony\Component\Validator\Constraints\IsTrue(message: "Game number must be between 1 and 3 (best-of-3)")]
    public function isGameNumberRangeValid(): bool
    {
        return ($this->getGameNumber() === null || ($this->getGameNumber() >= 1 && $this->getGameNumber() <= 3));
    }

    // ── Domain invariants (IMPLIES rules) ───────────────────────────────
    public function validateImplies(): void
    {
        if ($this->getTurnsPlayed() !== null && !(($this->getTurnsPlayed() === null || $this->getTurnsPlayed() > 0))) {
            throw new \DomainException('Turns played must be greater than zero');
        }
        if ($this->getDurationSeconds() !== null && !(($this->getDurationSeconds() === null || $this->getDurationSeconds() > 0))) {
            throw new \DomainException('Game duration must be greater than zero');
        }
    }

    // ── Business operations ──────────────────────────────────────────

    public function recordWinner($winnerSide): void
    {
        throw new \RuntimeException('record_winner not implemented');
    }

    public function durationMinutes(): void
    {
        throw new \RuntimeException('duration_minutes not implemented');
    }

}
